### ISSUE TITLE
`nodetool rebuild` operation causes the amount of disk space occupied by scylla to increase dramatically

### ISSUE THREAD

**Issue by ShlomiBalalis** [scylladb/scylladb#13580]
## Issue description

- [x] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

At `2023-03-22 04:33:34.256` a RepairStreamingErr nemesis has started with node 1 (3.252.128.60 | 10.4.1.126) as the target node. In the nemesis, we corrupt the node data by deleting 5 random sstable files in the target node, and execute a repair. Once the repair starts, we interrupt it by rebooting the node. Afterwards, we execute a rebuild.
Once the rebuild has started, at `2023-03-22 04:39:20,629`, node 1's occupied disk space has increased dramatically until the end of the rebuild, at `2023-03-22 07:05:45,165`, at which point it has increased from 1.2 TiB to over 2 TiB.
It also visible in the nodetool status output:
```
< t:2023-03-22 04:01:40,950 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.2.213  1018.71 GB  256          ?       b232cbf4-e63b-4119-a1a3-e779387eb9cf  1a
< t:2023-03-22 04:01:40,955 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.1.124  883.38 GB  256          ?       44674b58-acc2-4bbd-b854-d5d03bcef744  1a
< t:2023-03-22 04:01:40,955 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.1.126  1021.39 GB  256          ?       5a1cfde3-e243-4d03-83ce-888c109573e8  1a
< t:2023-03-22 04:01:40,962 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.3.73   1.07 TB    256          ?       0018ef85-105f-4d22-a241-e61ae1123dc6  1a

...
< t:2023-03-22 07:05:55,151 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.2.213  1020.1 GB  256          ?       b232cbf4-e63b-4119-a1a3-e779387eb9cf  1a
< t:2023-03-22 07:05:55,151 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.1.124  874.35 GB  256          ?       44674b58-acc2-4bbd-b854-d5d03bcef744  1a
< t:2023-03-22 07:05:55,182 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.1.126  1.78 TB    256          ?       5a1cfde3-e243-4d03-83ce-888c109573e8  1a
< t:2023-03-22 07:05:55,182 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.3.73   1.08 TB    256          ?       0018ef85-105f-4d22-a241-e61ae1123dc6  1a
```
The occupied space has only returned to normal at around `2023-03-22 10:35`, three and a half hours later.

The same situation has occurred in another nemesis run with node 2 (34.241.232.156 | 10.4.3.73). After the node was corrupted, a rebuild operation has started at `2023-03-22 07:40:54,884` and lasted until `2023-03-22 10:07:52,181`, which increased the occupied space from 1.28 TiB to 2.38 TiB. It even continued to increase after the rebuild until it reached 2.74 TiB.
```
< t:2023-03-22 07:05:55,151 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.2.213  1020.1 GB  256          ?       b232cbf4-e63b-4119-a1a3-e779387eb9cf  1a
< t:2023-03-22 07:05:55,151 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.1.124  874.35 GB  256          ?       44674b58-acc2-4bbd-b854-d5d03bcef744  1a
< t:2023-03-22 07:05:55,182 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.1.126  1.78 TB    256          ?       5a1cfde3-e243-4d03-83ce-888c109573e8  1a
< t:2023-03-22 07:05:55,182 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.3.73   1.08 TB    256          ?       0018ef85-105f-4d22-a241-e61ae1123dc6  1a
...
< t:2023-03-22 10:07:59,877 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.2.213  1020.47 GB  256          ?       b232cbf4-e63b-4119-a1a3-e779387eb9cf  1a
< t:2023-03-22 10:07:59,913 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.1.124  876.4 GB   256          ?       44674b58-acc2-4bbd-b854-d5d03bcef744  1a
< t:2023-03-22 10:08:00,136 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.1.126  1.05 TB    256          ?       5a1cfde3-e243-4d03-83ce-888c109573e8  1a
< t:2023-03-22 10:08:00,222 f:base.py         l:228  c:RemoteLibSSH2CmdRunner p:DEBUG > UN  10.4.3.73   1.97 TB    256          ?       0018ef85-105f-4d22-a241-e61ae1123dc6  1a
```

The problem became even more severe when a manager backup ran right after the rebuild. Since the manager creates a snapshot for all of the nodes' data, node 2 ran out of disk space and scylla has crashed as a result.

![Screenshot from 2023-04-19 18-50-37](https://user-images.githubusercontent.com/40186893/233131004-72289335-6974-4eec-bbf1-a138f4261c79.png)

## Impact

Could lead to any node that is being rebuilt to running out of disk space, and thus shutting down.

## How frequently does it reproduce?

In both of the rebuild operations in the run, the load in the rebuilt node has doubled.

## Installation details

Kernel Version: 5.15.0-1030-aws
Scylla version (or git commit hash): `5.2.0~rc2-20230228.908a82bea064` with build-id `c5e4affd90a7c49c0dcf1a00c3b8bc72a367aba2`

Cluster size: 4 nodes (i3.4xlarge)

Scylla Nodes used in this run:
- longevity-tls-1tb-7d-5-2-db-node-21572504-4 (54.171.106.131 | 10.4.2.213) (shards: 14)
- longevity-tls-1tb-7d-5-2-db-node-21572504-3 (34.244.158.156 | 10.4.1.124) (shards: 14)
- longevity-tls-1tb-7d-5-2-db-node-21572504-2 (34.241.232.156 | 10.4.3.73) (shards: 14)
- longevity-tls-1tb-7d-5-2-db-node-21572504-1 (3.252.128.60 | 10.4.1.126) (shards: 14)

OS / Image: `ami-07af3b969c1f912b8` (aws: eu-west-1)

Test: `longevity-1tb-7days-test`
Test id: `21572504-3888-45b9-b16d-a84cfd69e015`
Test name: `scylla-5.2/longevity/longevity-1tb-7days-test`
Test config file(s):

- [longevity-1TB-5days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/bd6309126a03e121e2fa5dd4b3fe07df0a326923/test-cases/longevity/longevity-1TB-5days-authorization-and-tls-ssl.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor 21572504-3888-45b9-b16d-a84cfd69e015`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=21572504-3888-45b9-b16d-a84cfd69e015)
- Show all stored logs command: `$ hydra investigate show-logs 21572504-3888-45b9-b16d-a84cfd69e015`

## Logs:
- **db-cluster-21572504.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/db-cluster-21572504.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/db-cluster-21572504.tar.gz)
- **email_data-21572504.json.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/email_data-21572504.json.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/email_data-21572504.json.tar.gz)
- **output-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/output-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/output-21572504.log.tar.gz)
- **debug-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/debug-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/debug-21572504.log.tar.gz)
- **events-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/events-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/events-21572504.log.tar.gz)
- **sct-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/sct-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/sct-21572504.log.tar.gz)
- **normal-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/normal-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/normal-21572504.log.tar.gz)
- **argus-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/argus-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/argus-21572504.log.tar.gz)
- **raw_events-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/raw_events-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/raw_events-21572504.log.tar.gz)
- **critical-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/critical-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/critical-21572504.log.tar.gz)
- **warning-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/warning-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/warning-21572504.log.tar.gz)
- **summary-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/summary-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/summary-21572504.log.tar.gz)
- **left_processes-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/left_processes-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/left_processes-21572504.log.tar.gz)
- **error-21572504.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/error-21572504.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/error-21572504.log.tar.gz)
- **monitor-set-21572504.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/monitor-set-21572504.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/monitor-set-21572504.tar.gz)
- **loader-set-21572504.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/loader-set-21572504.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/21572504-3888-45b9-b16d-a84cfd69e015/20230322_230049/loader-set-21572504.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-5.2/job/longevity/job/longevity-1tb-7days-test/4/)
</details>

**Comment by mykaul** [comment]
I'm not sure I understand the bug here. Especially until we understand what caused the disk space increase.

**Comment by DoronArazii** [comment]
@eliransin / @asias we need to understand if it's a blocker for 5.2

**Comment by mykaul** [comment]
Per my discussion with @ShlomiBalalis , we'll also need:
1. To understand what consumes the disk space
2. To understand if it's a regression - may have happened in previous releases and we did not notice thus far.

**Comment by asias** [comment]
The version (5.2.0~rc2-20230228.908a82bea064) you used in the test does not use RBNO rebuild. It is expected to stream all the data from the source node during rebuild. That's the reason you see the almost 2x disk usage growth.  

With RBNO rebuild,  it will only sync the data missing to the rebuild node. It is one of the benefits using RBNO rebuild. 

It is not a regression. 

 ```
    , enable_repair_based_node_ops(this, "enable_repair_based_node_ops", liveness::LiveUpdate, value_status::Used, true, "Set true to use enable repair based node operations instead of streaming based")
    , allowed_repair_based_node_ops(this, "allowed_repair_based_node_ops", liveness::LiveUpdate, value_status::Used, "replace", "A comma separated list of node operations which are allowed to enable repair based node operations. The operations can be bootstrap, replace, removenode, decommission and rebuild")
```

**Comment by mykaul** [comment]
Let's reproduce with an up-to-date version of 5.2.0.

**Comment by roydahan** [comment]
@ShlomiBalalis why do you claim it's a regression?
What evidence you have for previous releases?

**Comment by ShlomiBalalis** [comment]
> Let's reproduce with an up-to-date version of 5.2.0.

Here's a 5.2.0~rc4 run of the same scenario: https://jenkins.scylladb.com/job/scylla-5.2/job/SCT_Reproducers/job/longevity-1tb-7days-test_reproduce-13580/3/ (Been running since yesterday)

**Comment by mykaul** [comment]
We still don't know what occupies the extra space. @ShlomiBalalis - please run a before/after 'ls -lR' command on the disk, and we'll try to figure it out.

**Comment by ShlomiBalalis** [comment]
Running with 5.2.0~rc4, the issue does not reproduce:
![Screenshot from 2023-04-27 17-26-18](https://user-images.githubusercontent.com/40186893/234893744-a532f9d7-4d39-4bbc-bc9c-eceeb09b6031.png)

The rebuild started around `2022-04-27 00:00` on node `longevity-tls-1tb-7d-13580-re-db-node-6a3ff74d-2 (3.252.203.229 | 10.4.1.239)`. As seen in the screenshot, the disk usage of the node stayed consistent throughout the rebuild operation (so far at least, it has yet to end).

It did, however, cause 2 major changes:
First, the rebuild run time has increased enormously - in the rc2 run, the rebuild took only about a couple of hours. In the rc4 run, the rebuild has been running for over 17 hours and still did not complete.
Second, the node that was rebuilt was accompanied by several reactor stalls, some even nearly 2 seconds long.

I'll open a new issue for tracking.

## Installation details

Kernel Version: 5.19.0-1022-aws
Scylla version (or git commit hash): `5.2.0~rc4-20230402.d70751fee3f9` with build-id `80951fe7ff3c6e2c268211c71a9236071ac18a35`




Cluster size: 4 nodes (i3.4xlarge)

Scylla Nodes used in this run:
- longevity-tls-1tb-7d-13580-re-db-node-6a3ff74d-4 (54.247.46.231 | 10.4.2.48) (shards: 14)
- longevity-tls-1tb-7d-13580-re-db-node-6a3ff74d-3 (54.77.246.90 | 10.4.0.148) (shards: 14)
- longevity-tls-1tb-7d-13580-re-db-node-6a3ff74d-2 (3.252.203.229 | 10.4.1.239) (shards: 14)
- longevity-tls-1tb-7d-13580-re-db-node-6a3ff74d-1 (3.250.224.43 | 10.4.2.33) (shards: 14)


OS / Image: `ami-05ad5589d671f2b48` (aws: eu-west-1)

Test: `longevity-1tb-7days-test_reproduce-13580`
Test id: `6a3ff74d-38d8-4ad4-a61b-38180e48c9c8`
Test name: `scylla-5.2/SCT_Reproducers/longevity-1tb-7days-test_reproduce-13580`
Test config file(s):

- [longevity-1TB-5days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/bd6309126a03e121e2fa5dd4b3fe07df0a326923/test-cases/longevity/longevity-1TB-5days-authorization-and-tls-ssl.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor 6a3ff74d-38d8-4ad4-a61b-38180e48c9c8`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=6a3ff74d-38d8-4ad4-a61b-38180e48c9c8)
- Show all stored logs command: `$ hydra investigate show-logs 6a3ff74d-38d8-4ad4-a61b-38180e48c9c8`


## Logs:
Run is still in progress, will update this segments when the logs are collected.

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-5.2/job/SCT_Reproducers/job/longevity-1tb-7days-test_reproduce-13580/3/)
</details>

**Comment by fgelcer** [comment]
@DoronArazii , IMHO this one should make to 2023.1.0

**Comment by Mark-Gurevich** [comment]
This happened again on 2023.1.0-rc5

## Issue description

Disk space usage started to grow after restart on node longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-4 (10.4.3.122) at:

`2023-05-03T19:47:31+00:00 longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-4 !INFO | scylla[948]: [shard 0] init - Scylla version 2023.1.0~rc5-0.20230429.a47bcb26e42e initialization completed
`

Reached end of space at:
`< t:2023-05-04 00:48:29,403 f:base.py         l:325  c:sdcm.sct_events.base p:WARNING > [SCT internal warning] (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=24ccb25e-9b22-4402-8081-9225ef8c3aa2: type=NO_SPACE_ERROR regex=No space left on device line_number=934067 node=longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-4`

![disk_space_growth](https://github.com/scylladb/scylladb/assets/13746611/04d0b84e-64a8-4915-8754-0eabc98b7104)


## Installation details

Kernel Version: 5.15.0-1035-aws
Scylla version (or git commit hash): `2023.1.0~rc5-20230429.a47bcb26e42e` with build-id `d2644a8364f13d14d25be6b9d3c69f84612192bd`



Cluster size: 4 nodes (i3.4xlarge)

Scylla Nodes used in this run:
- longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-8 (34.243.186.105 | 10.4.0.247) (shards: 14)
- longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-7 (34.245.180.44 | 10.4.0.44) (shards: 14)
- longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-6 (54.170.150.29 | 10.4.0.147) (shards: 14)
- longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-5 (54.246.174.50 | 10.4.1.205) (shards: 14)
- longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-4 (34.242.251.208 | 10.4.3.122) (shards: 14)
- longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-3 (54.246.40.108 | 10.4.1.39) (shards: 14)
- longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-2 (54.246.18.178 | 10.4.2.146) (shards: 14)
- longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-1 (3.250.172.23 | 10.4.0.10) (shards: 14)


OS / Image: `ami-05e7801837cea47d9` (aws: eu-west-1)

Test: `longevity-1tb-7days-test`
Test id: `8ba46a7b-b854-4705-b4e3-23653a4f6ca9`
Test name: `enterprise-2023.1/disabled_raft/longevity-1tb-7days-test`
Test config file(s):

- [longevity-1TB-5days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/586748113f732e38b20e8718c6971652b3a892b4/test-cases/longevity/longevity-1TB-5days-authorization-and-tls-ssl.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor 8ba46a7b-b854-4705-b4e3-23653a4f6ca9`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=8ba46a7b-b854-4705-b4e3-23653a4f6ca9)
- Show all stored logs command: `$ hydra investigate show-logs 8ba46a7b-b854-4705-b4e3-23653a4f6ca9`


## Logs:
- **db-cluster-8ba46a7b.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/db-cluster-8ba46a7b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/db-cluster-8ba46a7b.tar.gz)
- **sct-runner-events-8ba46a7b.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/sct-runner-events-8ba46a7b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/sct-runner-events-8ba46a7b.tar.gz)
- **sct-8ba46a7b.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/sct-8ba46a7b.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/sct-8ba46a7b.log.tar.gz)
- **monitor-set-8ba46a7b.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/monitor-set-8ba46a7b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/monitor-set-8ba46a7b.tar.gz)
- **loader-set-8ba46a7b.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/loader-set-8ba46a7b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/8ba46a7b-b854-4705-b4e3-23653a4f6ca9/20230504_021755/loader-set-8ba46a7b.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/enterprise-2023.1/job/disabled_raft/job/longevity-1tb-7days-test/5/)
</details>

**Comment by mykaul** [comment]
@Mark-Gurevich - what was the disk space occupied with?

**Comment by asias** [comment]
> This happened again on 2023.1.0-rc5
> ## Issue description
> 
> Disk space usage started to grow after restart on node longevity-tls-1tb-7d-2023-1-db-node-8ba46a7b-4 (10.4.3.122) at:

I looked at the log. The disk grow was caused by the of the rebuild operation caused by the restart itself. 

The growth of disk space is expected. The rebuild uses the streaming based rebuild (without rbno rebuild).

The amount of used disk space could be doubled since rebuild will bring all the data it should have from peer nodes.  Compaction will reduce the amount disk. If compaction does not keep up, it will takes more time to reduce the disk usage.

It is a bad practice  to use rebuild if the node still have most of the data with the stream based rebuild. One should use repair instead in that case. 

 So what we could do:

1) Use rebuild with RBNO

or

2) Delete data before rebuild using rebuild without RBNO

or

3) Use repair instead of rebuild

**Comment by asias** [comment]
I summary, I do not think this is a real issue or a regression. The amount of total disk is up to 2X using rebuild without RBNO.  The peak total disk space is up to compaction speed. There is no guarantee it will below some number, say 1.2X or 1.5X.

**Comment by roydahan** [comment]
How do you explain the regression?
AFAIK the same test passes without reaching enospc in the last 5 years.

@ShlomiBalalis do we have a side by side rub with the same SCT branch for the two different versions?

**Comment by mykaul** [comment]
1. Future (>5.2) should use RBNO by default, no?
2. I'm not even sure - how would a user know that they should prefer (3) ('use repair instead of rebuild') ?
3. Would implementing https://github.com/scylladb/scylladb/issues/13308 may help?

**Comment by roydahan** [comment]
But 2023.1.0 isn't with RBNO and it's going to live with us for quite long.

**Comment by mykaul** [comment]
> But 2023.1.0 isn't with RBNO and it's going to live with us for quite long.

IIRC there's a (pending?) backport to enable some node ops? @bhalevy  ?

**Comment by asias** [comment]
> > > But 2023.1.0 isn't with RBNO and it's going to live with us for quite long.
> > 
> > 
> > IIRC there's a (pending?) backport to enable some node ops? @bhalevy ?
> 
> No, it doesn't enable any node ops, just backports some cleanups to node ops.
> 
> For the record I agree with the analysis and recommendations by @asias above in [#13580 (comment)](https://github.com/scylladb/scylladb/issues/13580#issuecomment-1550591697)
> 
> According to @ShlomiBalalis in [#13580 (comment)](https://github.com/scylladb/scylladb/issues/13580#issuecomment-1521750927) this isn't even a clear regression, but maybe an earlier trend that is somehow exacerbated in more recent releases. Not sure why, maybe we improved repair and we're able to repair more data before repair is interrupted and we start rebuild. All this doesn't matter much. It's a weird scenario and we have procedures to work around the problem. Like, why even run streaming-based rebuild after we already started repair?

Exactly. 

Rebuild is designed to work on an empty node.  See our docs. We have two usages of rebuild. Both of them are performed on empty nodes. 

Rebuild a Node After Losing the Data Volume
https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/rebuild-node.html

Adding a New Data Center Into an Existing ScyllaDB Cluster
https://docs.scylladb.com/stable/operating-scylla/procedures/cluster-management/add-dc-to-existing-dc.html

**Comment by asias** [comment]
> 1. Future (>5.2) should use RBNO by default, no?

Yes. 

>     2. I'm not even sure - how would a user know that they should prefer (3) ('use repair instead of rebuild') ?

I mentioned here: https://github.com/scylladb/scylladb/issues/13580#issuecomment-1552289765.

In short, use rebuild for empty nodes. Otherwise prefer repair. 

>     3. Would implementing [Repair should "compact" data on the receiving side before writing it to disk #13308](https://github.com/scylladb/scylladb/issues/13308) may help?

Nope. The complained case is the classic streaming which does not use repair. If repair was used in the first, we would not have the issue here.

**Comment by roydahan** [comment]
I doesn't matter what is ideal way to run rebuild.
If in the past rebuild of this scenario used to consume much less storage and now it isn't, it's a regression we at least need to be aware of it source and decide what to do with it.

**Comment by mykaul** [comment]
> I doesn't matter what is ideal way to run rebuild. If in the past rebuild of this scenario used to consume much less storage and now it isn't, it's a regression we at least need to be aware of it source and decide what to do with it.

Decide what to do == use repair in that scenario instead of rebuild. Sounds like a good path forward to me.

**Comment by roydahan** [comment]
Ok, so let's deprecate rebuild, because we don't know if it will consume more space than the disk has.

**Comment by bhalevy** [comment]
> Ok, so let's deprecate rebuild, because we don't know if it will consume more space than the disk has.

With RBNO enabled maybe rebuild just need to run repair automatically.
With streaming, as @asias said above, it's meant to run on an empty node,
so maybe it should verify that and refuse to run on node with data. Maybe, for backward compatibility we can implement  a `--force` or `--unsafe` flag to bypass the check.

**Comment by ShlomiBalalis** [comment]
I reran the longevity (strictly with the `RepairStreamingErr` nemesis) both with `5.1.10-20230508.37538f00f5ef` and `5.2.1-20230508.f1c45553bc29`, to see if there's a regression.

It appears that using rebuild in 5.2 does end up occupying significantly more disk space than it did in 5.1, and sometimes no extra space is occupied in the 5.1 runs:
**5.1**:
![Screenshot from 2023-05-31 14-39-31](https://github.com/scylladb/scylladb/assets/40186893/6c9ce501-6993-45e9-b67d-e0f834f58510)

**5.2**:
![Screenshot from 2023-05-31 14-39-55](https://github.com/scylladb/scylladb/assets/40186893/ccaea95e-b5f0-4f22-ae20-d38bb7bfe891)
(each red line is about the time the rebuild command was executed) 

We can also see that a rebuild takes longer to be executed in 5.1 than in 5.2:
**5.1**:
```
< t:2023-05-23 07:03:58,096 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-4 [54.195.189.66 | 10.4.3.186] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 12152.498511497 s
< t:2023-05-23 11:42:45,264 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-3 [34.241.193.143 | 10.4.2.130] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 14021.342599803003 s
< t:2023-05-23 17:06:08,721 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-1 [54.75.60.10 | 10.4.3.252] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 16715.310700563998 s
< t:2023-05-23 22:30:03,754 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-2 [3.252.160.199 | 10.4.1.204] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 16746.456436285996 s
< t:2023-05-24 04:58:02,060 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-1 [54.75.60.10 | 10.4.3.252] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 20792.820538763 s
```
**5.2**:
```
< t:2023-05-23 06:41:22,845 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-1 [52.213.125.8 | 10.4.3.145] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 10528.217435114002 s
< t:2023-05-23 09:53:36,166 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-3 [3.250.171.212 | 10.4.3.44] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 8974.841157579998 s
< t:2023-05-23 12:50:00,591 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-2 [52.210.182.227 | 10.4.2.51] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 8019.417990255002 s
< t:2023-05-23 16:26:24,327 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-3 [3.250.171.212 | 10.4.3.44] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 10395.898834656 s
< t:2023-05-23 19:49:52,989 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-2 [52.210.182.227 | 10.4.2.51] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 9654.268975421997 s
< t:2023-05-23 21:56:08,016 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-4 [54.154.181.102 | 10.4.1.186] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 5031.22848463901 s
< t:2023-05-24 01:45:13,862 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-4 [54.154.181.102 | 10.4.1.186] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 11217.964410711007 s
< t:2023-05-24 04:34:38,390 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-3 [3.250.171.212 | 10.4.3.44] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 7624.626630159008 s
< t:2023-05-24 06:57:36,370 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-1 [52.213.125.8 | 10.4.3.145] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 6049.4431830480025 s
```

## 5.1 run info:
## Installation details

Kernel Version: 5.15.0-1035-aws
Scylla version (or git commit hash): `5.1.10-20230508.37538f00f5ef` with build-id `a3fe8c264bbf449020c2465078c43f1b0e584e5d`




Cluster size: 4 nodes (i3.4xlarge)

Scylla Nodes used in this run:
- longevity-tls-1tb-7d-13580-Re-db-node-ed221584-4 (54.195.189.66 | 10.4.3.186) (shards: 14)
- longevity-tls-1tb-7d-13580-Re-db-node-ed221584-3 (34.241.193.143 | 10.4.2.130) (shards: 14)
- longevity-tls-1tb-7d-13580-Re-db-node-ed221584-2 (3.252.160.199 | 10.4.1.204) (shards: 14)
- longevity-tls-1tb-7d-13580-Re-db-node-ed221584-1 (54.75.60.10 | 10.4.3.252) (shards: 14)


OS / Image: `ami-0883294ed70769ac2` (aws: eu-west-1)

Test: `longevity-1tb-7days-test_13580_check--regression`
Test id: `ed221584-b2e9-4a18-9224-e4d335f141dd`
Test name: `scylla-5.2/SCT_Reproducers/longevity-1tb-7days-test_13580_check--regression`
Test config file(s):

- [longevity-1TB-5days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/243f860c3011bcc97ee57e4a5ae4ffbb38649695/test-cases/longevity/longevity-1TB-5days-authorization-and-tls-ssl.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor ed221584-b2e9-4a18-9224-e4d335f141dd`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=ed221584-b2e9-4a18-9224-e4d335f141dd)
- Show all stored logs command: `$ hydra investigate show-logs ed221584-b2e9-4a18-9224-e4d335f141dd`


## Logs:
- **db-cluster-ed221584.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/db-cluster-ed221584.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/db-cluster-ed221584.tar.gz)
- **sct-runner-events-ed221584.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/sct-runner-events-ed221584.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/sct-runner-events-ed221584.tar.gz)
- **sct-ed221584.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/sct-ed221584.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/sct-ed221584.log.tar.gz)
- **monitor-set-ed221584.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/monitor-set-ed221584.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/monitor-set-ed221584.tar.gz)
- **loader-set-ed221584.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/loader-set-ed221584.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ed221584-b2e9-4a18-9224-e4d335f141dd/20230524_070024/loader-set-ed221584.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-5.2/job/SCT_Reproducers/job/longevity-1tb-7days-test_13580_check--regression/4/)
[Argus](https://argus.scylladb.com/test/e878873e-fbcb-4292-9dab-dece621945d6/runs?additionalRuns[]=ed221584-b2e9-4a18-9224-e4d335f141dd)
</details>

## 5.2 run info:
## Installation details

Kernel Version: 5.15.0-1035-aws
Scylla version (or git commit hash): `5.2.1-20230508.f1c45553bc29` with build-id `88ac66b1719cc7c5b7e982aa34ba5dc95909b84a`




Cluster size: 4 nodes (i3.4xlarge)

Scylla Nodes used in this run:
- longevity-tls-1tb-7d-13580-Re-db-node-84504067-4 (54.154.181.102 | 10.4.1.186) (shards: 14)
- longevity-tls-1tb-7d-13580-Re-db-node-84504067-3 (3.250.171.212 | 10.4.3.44) (shards: 14)
- longevity-tls-1tb-7d-13580-Re-db-node-84504067-2 (52.210.182.227 | 10.4.2.51) (shards: 14)
- longevity-tls-1tb-7d-13580-Re-db-node-84504067-1 (52.213.125.8 | 10.4.3.145) (shards: 14)


OS / Image: `ami-0aca21ab7c357a6f0` (aws: eu-west-1)

Test: `longevity-1tb-7days-test_13580_check--regression`
Test id: `84504067-8c76-4b5c-a6e5-f2276a1b597c`
Test name: `scylla-5.2/SCT_Reproducers/longevity-1tb-7days-test_13580_check--regression`
Test config file(s):

- [longevity-1TB-5days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/243f860c3011bcc97ee57e4a5ae4ffbb38649695/test-cases/longevity/longevity-1TB-5days-authorization-and-tls-ssl.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor 84504067-8c76-4b5c-a6e5-f2276a1b597c`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=84504067-8c76-4b5c-a6e5-f2276a1b597c)
- Show all stored logs command: `$ hydra investigate show-logs 84504067-8c76-4b5c-a6e5-f2276a1b597c`


## Logs:
- **db-cluster-84504067.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/db-cluster-84504067.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/db-cluster-84504067.tar.gz)
- **sct-runner-events-84504067.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/sct-runner-events-84504067.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/sct-runner-events-84504067.tar.gz)
- **sct-84504067.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/sct-84504067.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/sct-84504067.log.tar.gz)
- **monitor-set-84504067.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/monitor-set-84504067.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/monitor-set-84504067.tar.gz)
- **loader-set-84504067.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/loader-set-84504067.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/84504067-8c76-4b5c-a6e5-f2276a1b597c/20230524_070017/loader-set-84504067.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-5.2/job/SCT_Reproducers/job/longevity-1tb-7days-test_13580_check--regression/5/)
[Argus](https://argus.scylladb.com/test/e878873e-fbcb-4292-9dab-dece621945d6/runs?additionalRuns[]=84504067-8c76-4b5c-a6e5-f2276a1b597c)
</details>

**Comment by asias** [comment]
> I reran the longevity (strictly with the `RepairStreamingErr` nemesis) both with `5.1.10-20230508.37538f00f5ef` and `5.2.1-20230508.f1c45553bc29`, to see if there's a regression.
> 
> It appears that using rebuild in 5.2 does end up occupying significantly more disk space than it did in 5.1, and sometimes no extra space is occupied in the 5.1 runs: **5.1**: ![Screenshot from 2023-05-31 14-39-31](https://user-images.githubusercontent.com/40186893/242270109-6c9ce501-6993-45e9-b67d-e0f834f58510.png)
> 
> **5.2**: ![Screenshot from 2023-05-31 14-39-55](https://user-images.githubusercontent.com/40186893/242270137-ccaea95e-b5f0-4f22-ae20-d38bb7bfe891.png) (each red line is about the time the rebuild command was executed)
> 
> We can also see that a rebuild takes longer to be executed in 5.1 than in 5.2: **5.1**:
> 
> ```
> < t:2023-05-23 07:03:58,096 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-4 [54.195.189.66 | 10.4.3.186] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 12152.498511497 s
> < t:2023-05-23 11:42:45,264 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-3 [34.241.193.143 | 10.4.2.130] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 14021.342599803003 s
> < t:2023-05-23 17:06:08,721 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-1 [54.75.60.10 | 10.4.3.252] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 16715.310700563998 s
> < t:2023-05-23 22:30:03,754 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-2 [3.252.160.199 | 10.4.1.204] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 16746.456436285996 s
> < t:2023-05-24 04:58:02,060 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-ed221584-1 [54.75.60.10 | 10.4.3.252] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 20792.820538763 s
> ```
> 
> **5.2**:
> 
> ```
> < t:2023-05-23 06:41:22,845 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-1 [52.213.125.8 | 10.4.3.145] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 10528.217435114002 s
> < t:2023-05-23 09:53:36,166 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-3 [3.250.171.212 | 10.4.3.44] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 8974.841157579998 s
> < t:2023-05-23 12:50:00,591 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-2 [52.210.182.227 | 10.4.2.51] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 8019.417990255002 s
> < t:2023-05-23 16:26:24,327 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-3 [3.250.171.212 | 10.4.3.44] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 10395.898834656 s
> < t:2023-05-23 19:49:52,989 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-2 [52.210.182.227 | 10.4.2.51] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 9654.268975421997 s
> < t:2023-05-23 21:56:08,016 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-4 [54.154.181.102 | 10.4.1.186] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 5031.22848463901 s
> < t:2023-05-24 01:45:13,862 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-4 [54.154.181.102 | 10.4.1.186] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 11217.964410711007 s
> < t:2023-05-24 04:34:38,390 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-3 [3.250.171.212 | 10.4.3.44] (seed: False): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 7624.626630159008 s
> < t:2023-05-24 06:57:36,370 f:cluster.py      l:2587 c:sdcm.cluster_aws     p:DEBUG > Node longevity-tls-1tb-7d-13580-Re-db-node-84504067-1 [52.213.125.8 | 10.4.3.145] (seed: True): Command '/usr/bin/nodetool -u cassandra -pw 'cassandra'  rebuild ' duration -> 6049.4431830480025 s
> ```
> 
> ## 5.1 run info:
> ## Installation details
> 
> Kernel Version: 5.15.0-1035-aws Scylla version (or git commit hash): `5.1.10-20230508.37538f00f5ef` with build-id `a3fe8c264bbf449020c2465078c43f1b0e584e5d`
> 
> Cluster size: 4 nodes (i3.4xlarge)
> 
> Scylla Nodes used in this run:
> 
>     * longevity-tls-1tb-7d-13580-Re-db-node-ed221584-4 (54.195.189.66 | 10.4.3.186) (shards: 14)
> 
>     * longevity-tls-1tb-7d-13580-Re-db-node-ed221584-3 (34.241.193.143 | 10.4.2.130) (shards: 14)
> 
>     * longevity-tls-1tb-7d-13580-Re-db-node-ed221584-2 (3.252.160.199 | 10.4.1.204) (shards: 14)
> 
>     * longevity-tls-1tb-7d-13580-Re-db-node-ed221584-1 (54.75.60.10 | 10.4.3.252) (shards: 14)
> 
> 
> OS / Image: `ami-0883294ed70769ac2` (aws: eu-west-1)
> 
> Test: `longevity-1tb-7days-test_13580_check--regression` Test id: `ed221584-b2e9-4a18-9224-e4d335f141dd` Test name: `scylla-5.2/SCT_Reproducers/longevity-1tb-7days-test_13580_check--regression` Test config file(s):
> 
>     * [longevity-1TB-5days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/243f860c3011bcc97ee57e4a5ae4ffbb38649695/test-cases/longevity/longevity-1TB-5days-authorization-and-tls-ssl.yaml)
> 
> 
> Logs and commands
> ## 5.2 run info:
> ## Installation details
> 
> Kernel Version: 5.15.0-1035-aws Scylla version (or git commit hash): `5.2.1-20230508.f1c45553bc29` with build-id `88ac66b1719cc7c5b7e982aa34ba5dc95909b84a`
> 
> Cluster size: 4 nodes (i3.4xlarge)
> 
> Scylla Nodes used in this run:
> 
>     * longevity-tls-1tb-7d-13580-Re-db-node-84504067-4 (54.154.181.102 | 10.4.1.186) (shards: 14)
> 
>     * longevity-tls-1tb-7d-13580-Re-db-node-84504067-3 (3.250.171.212 | 10.4.3.44) (shards: 14)
> 
>     * longevity-tls-1tb-7d-13580-Re-db-node-84504067-2 (52.210.182.227 | 10.4.2.51) (shards: 14)
> 
>     * longevity-tls-1tb-7d-13580-Re-db-node-84504067-1 (52.213.125.8 | 10.4.3.145) (shards: 14)
> 
> 
> OS / Image: `ami-0aca21ab7c357a6f0` (aws: eu-west-1)
> 
> Test: `longevity-1tb-7days-test_13580_check--regression` Test id: `84504067-8c76-4b5c-a6e5-f2276a1b597c` Test name: `scylla-5.2/SCT_Reproducers/longevity-1tb-7days-test_13580_check--regression` Test config file(s):
> 
>     * [longevity-1TB-5days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/243f860c3011bcc97ee57e4a5ae4ffbb38649695/test-cases/longevity/longevity-1TB-5days-authorization-and-tls-ssl.yaml)
> 
> 
> Logs and commands


What you observed in the test above matches what is supposed to be perfectly. 

$ git log  37538f00f5ef..f1c45553bc29 --author=asias --oneline|grep streaming
b9e5e340aa streaming: Enable offstrategy for all classic streaming based node ops

5.2 enabled off strategy for the classic streaming.

So

1) scylla 5.1

- does not use off-strategy compaction for rebuild
- compaction happens immediately when a sstable is received from streaming
- compaction runs in the normal compaction group which has 1000 shares

2) scylla 5.2

- use off-strategy compaction for rebuild
- compaction happens when a given table is finished streaming
- compaction runs in streaming group which has 200 shares

In 5.1 compaction runs much more aggressively. That's the reason why we saw
less disk usage. Because compaction runs both earlier and more aggressively,
the duplicated data is removed by compaction faster. It is also the reason why
rebuild takes much longer in 5.1 because more cpu and disk was consumed by
compaction during. It compacts again and again once a new sstable is added. In
contrast, 5.2 only needs to compact it once. 

The aggressive and wasteful compaction behavior in 5.1 is exactly what we want to solve with off strategy. 
The trade off is that we would have a slower compaction to reduce the duplicated data compared to 5.1. 
But in practice, rebuild is not supposed to to have any duplicated data in the first place. 

In summary, I do not think this is a real regression. In 5.2 rebuild is faster and more efficient and having less
impact on the CQL workload.

**Comment by roydahan** [comment]
Now that make more sense and explains the change in behavior.
I think we can close it.

**Comment by fgelcer** [comment]
but what does that mean? do we need to increase the disk in the instances from 5.2, so we won't get unexpected ENOSPC?

**Comment by bhalevy** [comment]
@asias shouldn't streaming wait on off-strategy compaction? If we're running out of space maybe streaming is creating too long backlog for off-strategy compaction and we need some form of back pressure.
How about limiting the number of background off-strategy to N, even N=1 to start with.
When streaming is done with a table it can wait until the number of outstanding off-strategy jobs is less than N (e.g. by acquiring a unit from a semaphore), and then start a background off-strategy job that will hold the permit until done.  This way the temporary space amplification will be more limited.

**Comment by asias** [comment]
> but what does that mean? do we need to increase the disk in the instances from 5.2, so we won't get unexpected ENOSPC?

I think we can delete data before running rebuild to test how rebuild is supposed to be used.
